filament/src/MorphTargetBuffer.cpp

```diff
@@ -74,114 +77,102 @@ static inline size_t getHeight(size_t vertexCount) noexcept {
     return (vertexCount + MAX_MORPH_TARGET_BUFFER_WIDTH) / MAX_MORPH_TARGET_BUFFER_WIDTH;
 }
 
-static inline size_t getDepth(size_t targetCount) noexcept {
-    return targetCount * 2;
-}
+template<VertexAttribute A>
+inline size_t getSize(size_t vertexCount) noexcept;
 
-static inline size_t getSize(size_t vertexCount) noexcept {
-    const size_t stride = getWidth(vertexCount) * sizeof(float4);
+template<>
+inline size_t getSize<VertexAttribute::POSITION>(size_t vertexCount) noexcept {
+    const size_t stride = getWidth(vertexCount) * sizeof(float3);
+    const size_t height = getHeight(vertexCount);
     return Texture::PixelBufferDescriptor::computeDataSize(
-            Texture::PixelBufferDescriptor::PixelDataFormat::RGBA,
+            Texture::PixelBufferDescriptor::PixelDataFormat::RGB,
             Texture::PixelBufferDescriptor::PixelDataType::FLOAT,
-            stride, getHeight(vertexCount), 1);
-}
-
-static inline size_t getPositionIndex(size_t targetIndex) noexcept {
-    return targetIndex * 2 + 0;
+            stride, height, 1);
 }
 
-static inline size_t getTangentIndex(size_t targetIndex) noexcept {
-    return targetIndex * 2 + 1;
+template<>
+inline size_t getSize<VertexAttribute::TANGENTS>(size_t vertexCount) noexcept {
+    const size_t stride = getWidth(vertexCount) * sizeof(short4);
+    const size_t height = getHeight(vertexCount);
+    return Texture::PixelBufferDescriptor::computeDataSize(
+            Texture::PixelBufferDescriptor::PixelDataFormat::RG_INTEGER,
+            Texture::PixelBufferDescriptor::PixelDataType::UINT,
+            stride, height, 1);
 }
 
 FMorphTargetBuffer::FMorphTargetBuffer(FEngine& engine, const Builder& builder)
         : mSBuffer(PerRenderPrimitiveMorphingSib::SAMPLER_COUNT),
           mVertexCount(builder->mVertexCount),
           mCount(builder->mCount) {
     FEngine::DriverApi& driver = engine.getDriverApi();
-
     mSbHandle = driver.createSamplerGroup(PerRenderPrimitiveMorphingSib::SAMPLER_COUNT);
-    mTbHandle = driver.createTexture(SamplerType::SAMPLER_2D_ARRAY, 1, TextureFormat::RGBA32F, 1,
-            getWidth(mVertexCount), getHeight(mVertexCount), getDepth(mCount),
-            TextureUsage::DEFAULT);
+    mPbHandle = driver.createTexture(SamplerType::SAMPLER_2D_ARRAY, 1, TextureFormat::RGB32F, 1,
```

Comment:
See my previous comment. Is RGB32F natively supported? If not we should just request RGBA32F (also could we use RGBA16F instead?)

Reply:
No it's not natively supported. I will reverted it.

Reply:
When I have tested it RGBA16F isn't enough. This data is very small so the precision is not enough.

Reply:
@romainguy I'm wrong. RGBA16F is enough for positions of morph targets. So I change this format to RGBA16F. But it will fallback to slow copy pass. I will enhance it later.

Reply:
I have checked using RGBA16F instead of RGBA32F and it looks fine. I have changed it.